addons/website_sale/controllers/main.py

```diff
@@ -1504,27 +1505,33 @@ def _get_shop_payment_values(self, order, **kwargs):
             'landing_route': '/shop/payment/validate',
         }
         values = {**portal_page_values, **payment_form_values}
-        if request.website.enabled_delivery:
-            has_storable_products = any(
-                line.product_id.type in ['consu', 'product'] for line in order.order_line
-            )
-            if not order._get_delivery_methods() and has_storable_products:
-                values['errors'].append((
-                    _('Sorry, we are unable to ship your order'),
-                    _('No shipping method is available for your current order and shipping address.'
-                    ' Please contact us for more information.')
-                ))
-
-            if has_storable_products:
-                if order.carrier_id and not order.delivery_rating_success:
-                    order._remove_delivery_line()
-                values['deliveries'] = order._get_delivery_methods().sudo()
-
-            values['delivery_has_storable'] = has_storable_products
-            values['delivery_action_id'] = request.env.ref(
-                'delivery.action_delivery_carrier_form'
-            ).id
+        values = self._get_accepted_carriers(order, values, False)
+        return values
 
+    def _get_accepted_carriers(self, order, values, express_payment, **kwargs):
+        if order.carrier_id and not order.delivery_rating_success:
+            order._remove_delivery_line()
+
+        delivery_carriers = order._get_delivery_methods()
+        accepted_carriers = []
+        for carrier in delivery_carriers:
+            if carrier.delivery_type == 'base_on_rule':
+                try:
+                    carrier._get_price_available(order)
+                    accepted_carriers.append(carrier)
+                except Exception:
```

Comment:
```suggestion
                except UserError:
```
Maybe it would be better to add a method/field in the carrier to properly assess that the carrier' price rule is applicable or not (what you do here).

Reply:
Hello @vava-odoo ! Thanks for the review ! Could you elaborate here what you mean ? I am not sure to clearly see you point of view :) The rest of the requested changes have been done :) Have a good day !

Reply:
Hello.
What I mean is that IMO this `try/except` part should be a proper method. Either `_get_price_available` should return `False` instead of a `UserError` (but this would surely break some flows), either you create a new method in delivery, something like `get_available_carriers()`, that would filter the carriers depending on the context (with an override in website_sale (?) if the price rule does not apply to the cart/order).
I'm not sure for this part, so I'd rather wait for the input of @Feyensv before changing this. It's not clean, but it works, so maybe we'll do it later in a refactor.